File: dce/day8/day8.hs
import Data.List
import Data.Maybe
import System.IO

splitInput :: [String] -> [String] -> ([String], [String])
splitInput _ [] = (["what"], ["why"])
splitInput inputs ("|" : outputs) = (inputs, outputs)
splitInput inputs (s : ss) = splitInput (inputs ++ [s]) ss

identifiableDigit :: String -> Bool
identifiableDigit s =
    let len = length s in
    len == 2 || len == 3 || len == 4 || len == 7

segments :: [String] -> [String]
segments digits = 
    let len i s = length s == i in
    let matching fn = head (filter fn digits) in

    let one = matching (len 2) in
    let seven = matching (len 3) in
    let four = matching (len 4) in
    let eight = matching (len 7) in

    let three = matching (\s -> len 5 s && len 0 (one \\ s)) in
    let five = matching (\s -> len 5 s && s /= three && len 1 (four \\ s)) in
    let two = matching (\s -> len 5 s && s /= three && s /= five) in

    let nine = matching (\s -> len 6 s && len 0 (three \\ s)) in
    let zero = matching (\s -> len 6 s && s /= nine && len 0 (seven \\ s)) in
    let six = matching (\s -> len 6 s && s /= nine && s /= zero) in

    map sort [zero, one, two, three, four, five, six, seven, eight, nine]

valueForRow :: ([String], [String]) -> Int
valueForRow (inputs, outputs) =
    let segs = segments (inputs ++ outputs) in
    let digits = map (\s -> fromJust $ elemIndex (sort s) segs) outputs in
    foldl (\sum i -> sum * 10 + i) 0 digits

main :: IO ()
main = do
    input <- readFile "./input.txt"
    let digits = map (splitInput [] . words) (lines input)
    let outputs = map snd digits

    let count = sum (map (length . filter identifiableDigit) outputs)

    putStrLn ("Part 1: " ++ (show count))

    let values = map valueForRow digits

    putStrLn ("Part 2: " ++ (show (sum values)))

File: noahover/lib/day8.ex
defmodule Day8 do
  @moduledoc """
  Advent of Code 2021 Day 8
  """

  def parse_input(input) do
    input
    |> String.split("\n", trim: true)
    |> Enum.map(fn x ->
      x
      |> String.split(" | ", trim: true)
      |> Enum.map(fn y ->
        y
        |> String.split(" ", trim: true)
        |> Enum.map(fn z ->
          z
          |> String.graphemes()
          |> Enum.sort()
          |> Enum.join()
        end)
      end)
      |> List.to_tuple()
    end)
  end

  @doc """
  Day 8 Part 1

  Find the number of unique lengths of the output strings
  """
  def unique_lengths(rows) do
    rows
    |> count_unique()
  end

  defp count_unique([]), do: 0
  defp count_unique([{_inputs, outputs} | tail]) do
    count = Enum.count(outputs, fn x ->
      len = String.length(x)
      len == 2 or len == 3 or len == 4 or len == 7
    end)
    count + count_unique(tail)
  end

  @doc """
  Day 8 Part 2

  Sum the decoded numbers
  """
  def sum_decoded(rows) do
    rows
    |> Enum.map(fn x -> decode(x) end)
    |> Enum.sum()
  end

  defp decode({inputs, outputs}) do
    map = map_inputs(inputs)

    outputs
    |> Enum.map(fn x -> Map.get(map, x) end)
    |> Enum.join()
    |> String.to_integer()
  end

  defp map_inputs(inputs) do
    one = Enum.find(inputs, fn x -> String.length(x) == 2 end)
    inputs = List.delete(inputs, one)
    four = Enum.find(inputs, fn x -> String.length(x) == 4 end)
    inputs = List.delete(inputs, four)
    seven = Enum.find(inputs, fn x -> String.length(x) == 3 end)
    inputs = List.delete(inputs, seven)
    eight = Enum.find(inputs, fn x -> String.length(x) == 7 end)
    inputs = List.delete(inputs, eight)
    nine = Enum.find(inputs, fn x -> String.length(x) == 6 and sublist?(x, four) end)
    inputs = List.delete(inputs, nine)
    zero = Enum.find(inputs, fn x -> String.length(x) == 6 and sublist?(x, one) end)
    inputs = List.delete(inputs, zero)
    six = Enum.find(inputs, fn x -> String.length(x) == 6 end)
    inputs = List.delete(inputs, six)
    three = Enum.find(inputs, fn x -> sublist?(x, one) end)
    inputs = List.delete(inputs, three)
    five = Enum.find(inputs, fn x -> sublist?(six, x) end)
    inputs = List.delete(inputs, five)
    two = Enum.at(inputs, 0)

    %{}
    |> Map.put(zero, "0")
    |> Map.put(one, "1")
    |> Map.put(two, "2")
    |> Map.put(three, "3")
    |> Map.put(four, "4")
    |> Map.put(five, "5")
    |> Map.put(six, "6")
    |> Map.put(seven, "7")
    |> Map.put(eight, "8")
    |> Map.put(nine, "9")
  end

  defp sublist?(str1, str2) do
    sublist_loop(String.graphemes(str1), String.graphemes(str2))
  end

  defp sublist_loop(_list1, []), do: true
  defp sublist_loop(list1, [head | tail]) do
    Enum.member?(list1, head) and sublist_loop(list1, tail)
  end
end
